### ISSUE TITLE
Can't force rebuilds of non-local packages

### ISSUE THREAD

**Issue by Blaisorblade** [commercialhaskell/stack#2818]
https://github.com/commercialhaskell/stack/issues/2805#issuecomment-263531315 reminded me of an issue I had recently but hadn't reported.

TL;DR. Sometimes you need to uninstall packages because they break somehow (because bugs) and you need to force a rebuild. I am 100% fine with "force a rebuild" if that's easier to implement. In fact, it'd be enough to extend `--force-dirty` (and its docs) to this use case.

```
$ stack install ghc-paths
Run from outside a project, using implicit global project config
Using resolver: lts-6.6 from implicit global project's config file: /Users/pgiarrusso/.stack/global/stack.yaml
$ stack install ghc-paths --force-dirty
Run from outside a project, using implicit global project config
Using resolver: lts-6.6 from implicit global project's config file: /Users/pgiarrusso/.stack/global/stack.yaml
$
```

Below I describe a crazy integration bug involving Homebrew, ghc-paths, and stack, as a concrete instance. This one might be Homebrew's fault but I'm not sure. I also realize I'm using a system GHC, but the plan is not to make that unsupported (since some users have valid use cases for that option and it makes sense to support them).

I use a system install of GHC 8.0.1, and installed it with Homebrew. Then, Homebrew updated it to version 8.0.1_1 (and then 8.0.1_2), which means some files were moved (they're now in `/usr/local/Cellar/ghc/8.0.1_2` rather than `/usr/local/Cellar/ghc/8.0.1`). But the GHC binary is still on the PATH, so 99% of packages don't care.

But I use intero. Which depends on ghc-paths. Which figures out those paths during the build, apparently—so they became invalid, and the Homebrew update broke intero.
So I had to uninstall ghc-paths to rebuild it. That was very challenging: I had to use `stack exec ghc-pkg unregister` (or some variation), but that wasn't enough because a new install fetched the compiled ghc-paths from some earlier snapshot. After a few tries, looking at verbose logs to figure out the involved earlier snapshot, I managed a clean reinstall. I sometimes removed the existing files by hand (after listing them with `stack exec ghc-pkg list`).

Yes, I know I could have nuked the entire cache, but many of you would probably find that annoying. Most users would only have that option, and I'm not 100% sure I don't have some silent corruption in the snapshot (though everything seems to work as expected).

I'm using stack 1.2.0 if that matters.

**Comment by kayhide** [comment]
Same problem for me.

My case is `doctest` which is also using `ghc-paths`.
After updating the homebrew-managed ghc, `doctest` complains:

```
% stack exec -- doctest
doctest: /usr/local/Cellar/ghc/8.0.1/bin/ghc-8.0.1: fileAccess: does not exist (No such file or directory)
```

Looking into the code of `ghc-paths`, it refers to envs which probably reflects the states of the local environment at the compile time.
https://github.com/simonmar/ghc-paths/blob/master/GHC/Paths.hs

Stack version is:
```
% stack --version
Version 1.3.0 x86_64 hpack-0.15.0
```

**Comment by Blaisorblade** [comment]
@kayhide thanks for confirming. Right now, as a workaround, I'm afraid the quickest and easiest thing is to remove your stack snapshots (while preserving your config); I don't have reliable step-by-step instructions for the alternative I did.
And I'm sure we don't need to get extra information from your broken state, since this is simply a missing feature.

Finally, a stack-managed GHC would avoid the problem, in case that's an option for you (it isn't for me).

**Comment by kayhide** [comment]
@Blaisorblade I removed snapshots and confirmed that `doctest` worked as expected.
Thanks for your quick replay :)

**Comment by mgsloan** [comment]
Summary of the issue: `stack exec -- ghc-pkg unregister X` is insufficient to force reinstall if `X` can be reused from a different snapshot DB.

I don't think it makes sense to change the meaning of `--force-dirty`, because this flag is used to override the file dirtiness check for local packages.  How about adding a `--reinstall` flag?  For target packages, this would cause stack to unregister the existing package and install a new version.

**Comment by Blaisorblade** [comment]
--reinstall sounds very fine; --force-dirty was the only remotely related
existing option.

-- 
Paolo Giarrusso
From smartphone, sorry for typos or excessive brevity

Il 17 dic 2016 23:27, "Michael Sloan" <notifications@github.com> ha scritto:

Summary of the issue: stack exec -- ghc-pkg unregister X is insufficient to
force reinstall if X can be reused from a different snapshot DB.

I don't think it makes sense to change the meaning of --force-dirty,
because this flag is used to override the file dirtiness check for local
packages. How about adding a --reinstall flag? For target packages, this
would cause stack to unregister the existing package and install a new
version.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/commercialhaskell/stack/issues/2818#issuecomment-267791184>,
or mute the thread

**Comment by dkasak** [comment]
Is it still the plan to implement this? I've been bitten by this so many times and it's always cumbersome to force stack to rebuild.

**Comment by Blaisorblade** [comment]
Based on @mgsloan's message, I've taken the liberty of adding it to the P2 milestone, so hopefully somebody picks it up. I sadly don't have time to do this myself but it sounds comparatively easy? As usual, time's the scarce resource.
